gui/default/syncthing/device/editDeviceModalView.html

```diff
@@ -10,19 +10,16 @@
         <div id="device-general" class="tab-pane in active">
           <div ng-if="!editingDeviceDefaults()" class="form-group" ng-class="{'has-error': deviceEditor.deviceID.$invalid && deviceEditor.deviceID.$dirty}" ng-init="loadFormIntoScope(deviceEditor)">
             <label translate for="deviceID">Device ID</label>
-            <div ng-if="!editingDeviceExisting()">
-              <div class="input-group">
-                <input name="deviceID" id="deviceID" class="form-control text-monospace" type="text" ng-model="currentDevice.deviceID" required="" valid-deviceid list="discovery-list" aria-required="true" />
-                <div class="input-group-btn">
-                  <button type="button" class="btn btn-default" data-toggle="modal" data-target="#idqr" ng-disabled="!deviceEditor.deviceID.$valid">
-                    <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
-                  </button>
-                </div>
+            <div class="input-group">
+              <input name="deviceID" id="deviceID" class="form-control text-monospace" type="text" ng-model="currentDevice.deviceID" required="" valid-deviceid list="discovery-list" aria-required="true" ng-readonly="currentDevice._editing !== 'add'" />
+              <div class="input-group-btn">
+                <button type="button" class="btn btn-default" data-toggle="modal" data-target="#idqr" ng-disabled="!deviceEditor.deviceID.$valid">
+                  <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
+                </button>
               </div>
-              <datalist id="discovery-list">
```

Comment:
Yes, I explained this above (and in the commit message).

Does this element serve any purpose other than to list the exact same set of device IDs as already listed below under the "nearby devices"?

Reply:
Sorry I missed the commit message.  A real pity that GitHub is so focused on PR conversations that most of commit messages are ususally not shown.  I use them quite extensively as well to document code changes.

Back to the matter: The datalist helps when navigating with the keyboard for example.  Start typing some characters inside the field, then rely on "auto-complete" behaviour.  At least there is no reason to remove the feature.

Reply:
> Back to the matter: The datalist helps when navigating with the keyboard for example. Start typing some characters inside the field, then rely on "auto-complete" behaviour. At least there is no reason to remove the feature.

To be frank, as it is right now, I personally find the datalist very annoying. It pops up as soon as you click on the device id input field, and while doing so, it also covers the actually useful list of nearby devices in the process.

![image](https://user-images.githubusercontent.com/5626656/149639606-4bea80e3-363c-4b14-bc16-dd7630e7454b.png)

The device IDs have their last part cut-off too, which makes it look kind of broken. It does help with keyboard autocomplete (although, at the same time, it's also very easy to access the nearby devices list through the TAB key), but how about reducing the listed entries to just one, so that autocomplete still works, but the list doesn't hover anything useful below?

Reply:
Note that this could look completely different depending on which browser.  Mobile devices or screen readers may make a totally different use of the data provided.  The app just says: "This may be meaningful inputs for the field", how the browser uses or renders it is up to them.

> To be frank, as it is right now, I personally find the datalist very annoying. It pops up as soon as you click on the device ID input field, and while doing so, it also covers the actually useful list of nearby devices in the process.

When someone explicitly clicks into the text entry field, I think they already decided to use the keyboard and not choose one from the list.  But the list may help to notice the string they're entering is actually among the discovered entries.  Again, think about a smartphone context.  The whole GUI might be invisible now because the virtual keyboard popped up to take 2/3 of the screen.

You may not need the feature, but there is no strong reason to remove it IMHO.

Reply:
> When someone explicitly clicks into the text entry field, I think they already decided to use the keyboard and not choose one from the list

Not really, I find myself often clicking on one of the nearby devices first, but then deciding not to use it and trying to remove the input ID manually. Doing so immediately triggers the datalist to pop up (which then results in the annoying covering of the nearby devices).

> Again, think about a smartphone context. The whole GUI might be invisible now because the virtual keyboard popped up to take 2/3 of the screen.

The situation on a smartphone is actually worse. This is a screenshot from an actual Android phone.

![image](https://user-images.githubusercontent.com/5626656/149640553-6ac37b66-0b5b-49de-b65d-0605fcd3be81.png)

The list is cut off even more, and the position is messed up too. Of course, there will be differences depending on the OS and the browser, but as it stands right now, I'd say it gets in the way much more than it actually helps.

It's also impossible to improve the situation by forcing the text to wrap or something, because browsers apparently still don't allow any styling of `datalist` and `option`. The current display is totally dependent on how each browser wishes to do it, with no control over it in the Web developer's hands.

I have just tested modifying the datalist to show only a single entry, but it still ended up covering the first ID in the nearby devices, so it's not really useful, at least with the current positioning.

Reply:
What's wrong with covering the nearby devices list when you decided not to need one of those IDs?

How (badly) browsers render it should not mean we don't hand out the information to e.g. screen readers.  I don't think trying to style the popup or limiting it to one entry (which one?) to be a sensible solution.

Reply:
> What's wrong with covering the nearby devices list when you decided not to need one of those IDs?

The problem is I haven't. I'm still undecided, but I've clicked on the input field first (maybe out of habit). Then I thought that maybe the device I want to add has already been listed by Syncthing automatically. However, now the nearby devices list is already covered by the other one.

Yes, I know I could then rely on the datalist instead to choose the device in question, but I personally find the whole thing confusing, because at this point there are basically two lists displayed on the screen, none of which is fully visible. The nearby devices list shows only the last parts of the device IDs, while the datalist, displayed on top of everything else, shows the device IDs cut-off.

Basically, with the two lists displayed at once, there is no place, where we can see the full device IDs.

> How (badly) browsers render it should not mean we don't hand out the information to e.g. screen readers.

While true, I'd personally be very surprised if the current UI was usable by an actual blind person with a screen reader. There is too much invalid HTML paired up with JS-generated content, and Bootstrap and Angular aren't known for being accessibility-friendly either.

Nevertheless, I wouldn't have any problem with the datalist if it didn't cover the other useful content (and, to a lesser extent, at least be able to display the device IDs in full).

Reply:
> However, now the nearby devices list is already covered by the other one.

Hit escape when you want to get rid of some value selection popup?  My point is that what covers what is out of our control here.  Giving the data to the browser to hopefully do something useful is the best we can do.

> Basically, with the two lists displayed at once, there is no place, where we can see the full device IDs.

If the one you want is actually one of the nearby devices, you only need the first (or last) few characters to determine which one.  Or the identicon added recently.  A truncated device ID is very much good enough for a list of max. 5 discovered devices.

Reply:
I don't think we're going to reach an agreement here. I still find the two lists displayed one on top of the other confusing. In the current state, it's difficult to focus on either of them. Yes, I know it's possible to press ESC or click somewhere on the screen to get rid of the pop-up. That's what I've actually been doing all the time so far. It still doesn't change the fact that the whole behaviour is very annoying (for me).

I may still remove this part of the commit just to make the whole PR more specific (and be more in line with the title). If there's no consensus regarding the datalist, I'm probably going to stick to getting rid of it in my personal builds only.

Reply:
I'm marking this one "resolved", as I've got rid of the datalist related change, but I'd also like to add that not only are the nearby devices covered by the pop-up, but also the warnings and the explanation are covered too.

Just have a look at the screenshots below.

Warnings covered:

![image](https://user-images.githubusercontent.com/5626656/149661243-68868dc0-23d3-46bb-b0f7-726517d59d90.png)
![image](https://user-images.githubusercontent.com/5626656/149661253-52098771-98a7-450b-abdf-6e061666bc82.png)

Explanation covered:

![image](https://user-images.githubusercontent.com/5626656/149661257-1e4acf17-6911-4330-98db-c29d426665bc.png)